Scripts/summaryStatistics/biophysics/growth_rate_analysis.py

```diff
@@ -0,0 +1,247 @@
+    @param  bin_radius  Thickness of band to average over (um)
+    @param  centerX     Center of colony (X-coord)
+    @param  centerY     Center of colony (Y-coord)
+    @return df          dataFrame containing: ['Radius'], ['Measurement'], ['bin']
+    """ 
+    # Assign bins to each r and measurement
+    bin_ranges = define_bin_ranges(bin_radius, r)
+    bin_indices = np.digitize(r, bin_ranges) #Allocate each r to a bin number; first bin = 1 #https://www.adamsmith.haus/python/answers/how-to-put-data-into-bins-using-numpy-in-python
+    
+    # Create a dataFrame of radial position, measurement, and bin index
+    df = pd.DataFrame()
+    df['Radius'] = r.tolist()
+    df['Growth Rate'] = measurement.tolist()
+    df['bin'] = bin_indices.tolist()
+    
+    return df
+    
+def get_binned_statistics(df):
+    """
+    Calculate the mean and standard deviation of binned data
+    
+    @param  df          dataFrame with a column ['bin']; the column contains integers
+    @return mean, std   Group dataFrames: mean and standard deviation of measurements (float)
+    """
+    # Group each cell by bin
+    df_groupby_bin = df.groupby(df['bin'])
+    
+    # Calculate statistics for all other measurements
+    mean = df_groupby_bin.mean()
+    std = df_groupby_bin.std()
+    
+    return mean, std
+    
+def filter_by_cell_age(cells, min_age=2):
+    """
+    Create a dict with cells that have age >= min_age
+    
+    @param  cells               cellStates dict
+    @param  min_age             minimum age to include cell in the filtered list; recommended age is 2
+    @return filtered_cellstates cellStates dict with cells filtered by age
+    """
+    filtered_cellstates = {}
+    for i, cell in cells.items():
+        if cell.cellAge >= min_age:
+            filtered_cellstates[i] = cell
+    
+    return filtered_cellstates
+    
+def get_all_centroids(cells):
+    """
+    Obtain centroids for all cells
+    
+    @param  cells           cellStates dict
+    @return all_centroids   nparray (n_cells, 2) of cell centroids
+    """
+    # Initialize storage variables
+    n_cells = len(cells.keys())
+    all_centroids = np.zeros((n_cells, 2))
+    
+    # Obtain centroids for all cells
+    for i, cell in enumerate(cells.values()):
+        all_centroids[i][0] = cell.pos[0]
+        all_centroids[i][1] = cell.pos[1]
+        
+    return all_centroids
+    
+def get_colony_centroid(cells):
+    '''
+    Fit an ellipse around colony using the welzl function
+    
+    @param  cells            cellStates dict
+    @return centerX, centerY colony centroid coordinates [x0, y0]
+    '''
+    # Get cell centroids
+    cell_centroids = get_all_centroids(cells)
+    
+    # Required check for larger colonies
+    if len(cell_centroids) >= 1000:
+        sys.setrecursionlimit(10000) # necessary to run welzl for colonies with n_cells > 1000  
+    
+    # Fit ellipse     
+    center, major, minor, theta = welzl(cell_centroids)  
```

Comment:
We need to fit an ellipse around the bacteria to calculate the aspect ratio. Therefore, using the center of the bacteria to fit the ellipse is not correct, and the endpoints of the bacteria must be used.

Reply:
See previous comment about this. For this summary statistic, only the centroid is important. I will again argue that for a reasonably large colony (maybe more than 10-20 cells), the center of mass of the colony will not change much if you fit an ellipse around centroids vs. end points.